### PR TITLE
Add basic auth filter

### PR DESCRIPTION
**Pull request by bianpengyuan**
Signed-off-by: Pengyuan Bian <bianpengyuan@google.com>

### CODE REVIEW

**Comment by bianpengyuan** [comment]
I ported the basic auth filter from istio/proxy repo to here: https://github.com/istio/proxy/tree/master/extensions/basic_auth. I also modified the code a bit and add unit tests with a mock host wasm context.

Let me know if you'd rather want me to check in a simple helloworld example with Wasm bazel stuff, so that we can shrink this PR to just basic auth.

Currently presubmit runs unit test and integration test. I will work on coverage report in later PRs.

File: .bazelrc [hunk 1]
```diff
@@ -1,14 +1,13 @@
...
-build --crosstool_top=//toolchain:emscripten
-
-# Use --cpu as a differentiator.
-build --cpu=wasm32
-
-# Use the default Bazel C++ toolchain to build the tools used during the
-# build.
-
-build --host_crosstool_top=@bazel_tools//tools/cpp:toolchain
+build --cxxopt=-std=c++17
```
Comment (kyessenov) [inline .bazelrc:1]: I don't see `cpu=wasm32` flag? Should we add a config for it?

Comment (kyessenov) [reply .bazelrc:1]: Ah, so we still need x86 for unit tests.

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+          if [[ "{$target}" == *":_wasm_"* ]]; then
+            # skip wasm transition target
+            continue
+          fi
+          "${GITHUB_WORKSPACE}/bin/bazel" build "${target}"
+        done
+
+    - name: Unit Test
+      run: |
+        "${GITHUB_WORKSPACE}/bin/bazel" test //...
```
Comment (kyessenov) [inline .github/workflows/test.yaml:56]: add `--test_output=errors`

File: WORKSPACE [hunk 3]
```diff
@@ -1,10 +1,81 @@
 workspace(name = "istio_ecosystem_wasm_extensions")
 
-load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive", "http_file")
+load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository")
 
+# reference the same commit as istio release-1.7
```
Comment (kyessenov) [inline WORKSPACE:6]: TODO: this needs to be turned into a bazel library (in istio/proxy for now I guess), so it is not duplicated.

File: bazel/wasm/wasm.bzl [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# Forked from https://github.com/istio/envoy/blob/release-1.7/bazel/wasm/wasm.bzl
```
Comment (kyessenov) [inline bazel/wasm/wasm.bzl:1]: This should be a bazel library, too.

File: WORKSPACE (hunk 3 above)
Comment (bianpengyuan) [reply WORKSPACE:6]: Added a TODO for this.

File: .bazelrc (hunk 1 above)
Comment (bianpengyuan) [reply .bazelrc:1]: Yeah, the CPU option is added in the wasm binary transition rule: https://github.com/istio-ecosystem/wasm-extensions/pull/4/files#diff-1313d11afd1f5647109144e5f0283fa9R8

**Review by kyessenov — APPROVED** [review]

**Pull request merged**